### PR TITLE
Add `dandi service-scripts update-dandiset-from-doi` command

### LINKED ISSUE #1280: service-scripts update-dandiset-from-doi
In our today's meeting attended by @waxlamp @bendichter @asmacdo and @weiglszonja we decided that https://github.com/dandi/dandi-archive/issues/1423 would be approached by first implementing needed logic in dandi-cli, at large reusing prototypes coded up in https://github.com/dandi/dandi-archive/issues/1423#issuecomment-1373745996 and https://github.com/dandi/dandi-archive/issues/1423#issuecomment-1408669132 .

In this issue we would **not** work to address https://github.com/dandi/dandi-archive/issues/772 , as to enhance authors' metadata with records from ORCID. For that we might later craft a dedicated `update-dandiset-contributors-from-orcid` or alike to be either chained or integrated with in `update-dandiset-from-doi`.

I think we need 

```shell
dandi service-scripts update-dandiset-from-doi -d|--dandiset DANDISET_ID[ --existing ask,overwrite,skip][ --fields contributor,name,description,relatedResource]  DOI
```

- decided to keep `-d|--dandiset DANDISET_ID` and option (not positional) so later if desired we could just pipe through stdin etc
- `--existing` is to instruct on how to behave while intending to place a value where there is a **different** value already. If "ask" - show existing value, offer a new one, ask it should be overwritten.  if "overwrite" -- just overwrite without interaction. if "skip" - skip
- `--fields` -- by default go through all listed below, but if option is specified (comma separated values?) - consider only those

The command will
- get metadata for DANDISET_ID
- request citation record from DOI
- based on information in DOI record
  - add "authors" to the `contributor:` list
    - while considering an author, check if there is already one which matches either by ORCID or if not present - name matches an already known contributor.  If so and not identical -- act based on `existing` value
    - ~~if corresponding author of the paper (if record provides that information) -- add with `dcite:ContactPerson` role,~~  for all - add as `dcite:Author`
    - see e.g. https://github.com/dandisets/000402/blob/b5e4ab7e991e1f91476da31c722be276ec08c8f3/dandiset.yaml#L121 for a sample of such automatically added entries and what default fields to take
  - use "title" for the "name:"
  - use "abstract" for the "description:"
  - add a record to `relatedResource:` list alike
 ```yaml
- identifier: {DOI}
  name: {title}
  relation: dcite:IsDescribedBy
  schemaKey: Resource
  url: https://doi.org/{DOI}
```
   - first check if there is one among possibly present and identify based on "identifier", act based on `--existing` 
- if anything was actually changed in the metadata record
  - add to `wasGeneratedBy:` a record like the one we are adding while extracting metadata for assets:

```json
        {
          "description": "Metadata ({','.join(modified_fields)}) was enhanced by DANDI cli from DOI {doi}",
          "endDate": "{now}",
          "id": "urn:uuid:{iirc random UUID}",
          "name": "Metadata extraction from DOI",
          "schemaKey": "Activity",
          "startDate": "{when started doing the mess}",
          "wasAssociatedWith": [
            {
              "identifier": "RRID:SCR_019009",
              "name": "DANDI Command Line Interface",
              "schemaKey": "Software",
              "url": "https://github.com/dandi/dandi-cli",
              "version": "{version}"
            }
          ]
        }
 ```
    which would then allow us to figure out dandisets which used this functionality.
   - ask to either show simple unified diff in rendering of the original and modified records
   - ask to either update dandiset record with this modified record, and if yes - do so

@bendichter @satra any other fields possibly to re-use or reference to from  dandiset?  Anything to add/remove from above?

### PR DESCRIPTION
**Pull request by jwodder**
Closes #1280.

TO DO:

- ~Detect whether a given author in the DOI data is a corresponding author and, if so, add the role "`dcite:ContactPerson`"~
- [x] Improve the log messages when adding a dict to a list field
    - [x] Improve the display of the new dict value
- [x] Add some testing
    - [x] Use vcr?
- [x] Update Sphinx docs

### CODE REVIEW

**Comment by jwodder** [comment]
@yarikoptic How exactly should this be tested?  Should the tests create a test DOI for a Dockerized Dandiset (Would that work?) and use that?  Should I use staging instead?  Either way, should the command gain a `--dandi-instance` option so that it can be used with the Dockerized or staging instance, or should there be some other CLI syntax?

File: dandi/cli/base.py [hunk 1]
```diff
@@ -30,6 +30,29 @@ def get_metavar(self, param):
...
+        if value is None or isinstance(value, set):
+            return value
+        selected = set()
+        for v in value.split(","):
+            if v == "all":
+                selected = self.values.copy()
+            elif v in self.values:
+                selected.add(v)
+            else:
+                self.fail(f"{v!r}: invalid value", param, ctx)
```
Comment (yarikoptic) [inline dandi/cli/base.py:49]: I see that it is just a moved code but may be while at it:

```suggestion
                self.fail(f"{v!r}: invalid value. Known are: {', '.join(self.values)}", param, ctx)
```
(would lack `all`) or

```suggestion
                self.fail(f"{v!r}: invalid value. Known are: {self.get_metavar(None)}", param, ctx)
```

to make it easier for people to enter correct values.   Someone fancy could have even used smth like https://github.com/datalad/datalad/blob/HEAD/datalad/utils.py#L2587 to assume that it was a typo and suggest close by choices

**Review by yarikoptic — CHANGES_REQUESTED** [review]
Looks awesome, thank you @jwodder!  But it seems to need a little more guarding.  I have tried on a sample invocation with some random DOI against 000029 dandiset, and it crashed

<details>
<summary>full protocol</summary> 

```shell
❯ dandi service-scripts update-dandiset-from-doi -d 000029 -e ask 10.7554/eLife.61834
2023-05-11 10:52:30,318 [ WARNING] A newer version (0.54.0) of dandi/dandi-cli is available. You are using 0.53.0+19.g2e3c6cce
2023-05-11 10:52:30,910 [    INFO] New value to add to Dandiset contributor field:

    affiliation:
    - name:
        name: Department of Biosystems Science and Engineering, ETH Zurich, Zürich, Switzerland
    - name:
        name: Centre for Integrative Neuroplasticity (CINPLA), University of Oslo, Oslo,
          Norway
    identifier: 0000-0003-3661-527X
    includeInCitation: true
    name: Buccino, Alessio P
    roleName:
    - dcite:Author
    schemaKey: Person


Add value to Dandiset contributor field? [Y/n] Y
2023-05-11 10:52:45,283 [    INFO] Adding value to Dandiset contributor field
2023-05-11 10:52:45,284 [    INFO] New value to add to Dandiset contributor field:

    affiliation:
    - name:
        name: School of Informatics, University of Edinburgh, Edinburgh, United Kingdom
    identifier: 0000-0002-2023-1653
    includeInCitation: true
    name: Hurwitz, Cole L
    roleName:
    - dcite:Author
    schemaKey: Person


Add value to Dandiset contributor field? [Y/n] Y
2023-05-11 10:52:52,954 [    INFO] Adding value to Dandiset contributor field
2023-05-11 10:52:52,955 [    INFO] New value to add to Dandiset contributor field:

    affiliation:
    - name:
        name: Centre de Recherche en Neuroscience de Lyon, CNRS, Lyon, France
    identifier: 0000-0001-6389-9779
    includeInCitation: true
    name: Garcia, Samuel
    roleName:
    - dcite:Author
    schemaKey: Person


Add value to Dandiset contributor field? [Y/n] Y
2023-05-11 10:52:57,691 [    INFO] Adding value to Dandiset contributor field
2023-05-11 10:52:57,693 [    INFO] New value to add to Dandiset contributor field:

    affiliation:
    - name:
        name: Flatiron Institute, New York, United States
    identifier: 0000-0002-5286-4375
    includeInCitation: true
    name: Magland, Jeremy
    roleName:
    - dcite:Author
    schemaKey: Person


Add value to Dandiset contributor field? [Y/n] Y
2023-05-11 10:53:01,212 [    INFO] Adding value to Dandiset contributor field
2023-05-11 10:53:01,214 [    INFO] New value to add to Dandiset contributor field:

    affiliation:
    - name:
        name: Allen Institute for Brain Science, Seattle, United States
    identifier: 0000-0002-7736-4844
    includeInCitation: true
    name: Siegle, Joshua H
    roleName:
    - dcite:Author
    schemaKey: Person


Add value to Dandiset contributor field? [Y/n] Y
2023-05-11 10:53:04,723 [    INFO] Adding value to Dandiset contributor field
Traceback (most recent call last):
  File "/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/bin/dandi", line 8, in <module>
    sys.exit(main())
             ^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/lib/python3.11/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/lib/python3.11/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
         ^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/lib/python3.11/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/lib/python3.11/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/venvs/dev3.11/lib/python3.11/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/dandi/cli/cmd_service_scripts.py", line 223, in update_dandiset_from_doi
    if add_dict_to_list_field(
       ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/dandi/cli/cmd_service_scripts.py", line 362, in add_dict_to_list_field
    matches = [item for item in metadata[key] if eqtest(newvalue, item)]  # type: ignore[arg-type]
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/dandi/cli/cmd_service_scripts.py", line 362, in <listcomp>
    matches = [item for item in metadata[key] if eqtest(newvalue, item)]  # type: ignore[arg-type]
                                                 ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/yoh/proj/dandi/dandi-cli-master/dandi/cli/cmd_service_scripts.py", line 429, in eq_authors
    return bool(author1["name"] == author2["name"])
                                   ~~~~~~~^^^^^^^^
KeyError: 'name'

```
</details>

**Comment by jwodder** [comment]
@yarikoptic 

> I have tried on a sample invocation with some random DOI against 000029 dandiset, and it crashed

That happened because one of the `contributor`s in the 000029 draft metadata is just an empty dict. I feel like the Dandi Archive server should reject such metadata.

File: dandi/cli/base.py (hunk 1 above)
Comment (jwodder) [reply dandi/cli/base.py]: Done: 04fbfa1

**Comment by yarikoptic** [comment]
thank you @jwodder ! Let's put it "in production" and see if anyone manages to break it ;)

**Pull request merged**